src/query/plan/rewrite/join.hpp

```diff
@@ -0,0 +1,547 @@
+    MG_ASSERT(input);
+    if (prev_ops_.empty()) {
+      MG_ASSERT(!new_root_);
+      new_root_ = input;
+      return;
+    }
+    prev_ops_.back()->set_input(input);
+  }
+
+  void RewriteBranch(std::shared_ptr<LogicalOperator> *branch) {
+    JoinRewriter<TDbAccessor> rewriter(symbol_table_, ast_storage_, db_);
+    (*branch)->Accept(rewriter);
+    if (rewriter.new_root_) {
+      *branch = rewriter.new_root_;
+    }
+  }
+
+  std::unique_ptr<HashJoin> GenHashJoin(const Cartesian &cartesian) {
+    const auto &left_op = cartesian.left_op_;
+    const auto &left_symbols = cartesian.left_symbols_;
+    const auto &right_op = cartesian.right_op_;
+    const auto &right_symbols = cartesian.right_symbols_;
+
+    auto modified_symbols = cartesian.ModifiedSymbols(*symbol_table_);
+    modified_symbols.insert(modified_symbols.end(), cartesian.left_symbols_.begin(), cartesian.left_symbols_.end());
+
+    std::unordered_set<Symbol> bound_symbols(modified_symbols.begin(), modified_symbols.end());
+    auto are_bound = [&bound_symbols](const auto &used_symbols) {
+      for (const auto &used_symbol : used_symbols) {
+        if (!utils::Contains(bound_symbols, used_symbol)) {
+          return false;
+        }
+      }
+      return true;
+    };
+
+    for (const auto &filter : filters_) {
+      if (filter.type != FilterInfo::Type::Property) {
+        continue;
+      }
+
+      if (filter.property_filter->is_symbol_in_value_ || !are_bound(filter.used_symbols)) {
+        continue;
+      }
+
+      if (filter.property_filter->type_ != PropertyFilter::Type::EQUAL) {
+        continue;
+      }
+
+      if (filter.property_filter->value_->GetTypeInfo() != PropertyLookup::kType) {
+        continue;
+      }
+      auto rhs_lookup = static_cast<PropertyLookup *>(filter.property_filter->value_);
+
+      auto join_condition = static_cast<EqualOperator *>(filter.expression);
+      auto lhs_symbol = filter.property_filter->symbol_;
+      auto lhs_property = filter.property_filter->property_;
+      auto rhs_symbol = symbol_table_->at(*static_cast<Identifier *>(rhs_lookup->expression_));
+      auto rhs_property = rhs_lookup->property_;
+      filter_exprs_for_removal_.insert(filter.expression);
+      filters_.EraseFilter(filter);
+      return std::make_unique<HashJoin>(left_op, left_symbols, right_op, right_symbols, join_condition);
+    }
+
+    return nullptr;
+  }
+};
+
+}  // namespace impl
+
+template <class TDbAccessor>
+std::unique_ptr<LogicalOperator> RewriteWithJoinRewriter(std::unique_ptr<LogicalOperator> root_op,
+                                                         SymbolTable *symbol_table, AstStorage *ast_storage,
+                                                         TDbAccessor *db) {
+  impl::JoinRewriter<TDbAccessor> rewriter(symbol_table, ast_storage, db);
+  root_op->Accept(rewriter);
+  if (rewriter.new_root_) {
+    throw utils::NotYetImplemented("optimizing join");
```

Comment:
I would be more descriptive with message here. Can you elaborate what is not yet implemented here

Reply:
Changed (@Josipmrden: maybe you need to do the same in #1193 in index_lookup.hpp)